pyomo/contrib/mindtpy/iterate.py

```diff
@@ -23,54 +25,75 @@ def MindtPy_iteration_loop(solve_data, config):
+        if fix_nlp_result.solver.termination_condition is tc.optimal:
+            handle_NLP_subproblem_optimal(fix_nlp, solve_data, config)
+        elif fix_nlp_result.solver.termination_condition is tc.infeasible:
+            handle_NLP_subproblem_infeasible(fix_nlp, solve_data, config)
         else:
-            log = solve_data.UB_progress
-            sign_adjust = -1
-        # Maximum number of iterations in which the lower (optimistic)
-        # bound does not improve before switching to OA
-        max_nonimprove_iter = 5
-        making_progress = True
-        for i in range(1, max_nonimprove_iter + 1):
-            try:
-                if (sign_adjust * log[-i]
-                        <= (log[-i - 1] + progress_required)
-                        * sign_adjust):
-                    making_progress = False
-                else:
+            handle_NLP_subproblem_other_termination(fix_nlp, fix_nlp_result.solver.termination_condition,
+                                                    solve_data, config)
+        # Call the NLP post-solve callback
+        config.call_after_subproblem_solve(fix_nlp, solve_data)
+
+        if config.strategy == 'PSC':
+            # If the hybrid algorithm is not making progress, switch to OA.
+            progress_required = 1E-6
+            if main_objective.sense == minimize:
+                log = solve_data.LB_progress
+                sign_adjust = 1
+            else:
+                log = solve_data.UB_progress
+                sign_adjust = -1
+            # Maximum number of iterations in which the lower (optimistic)
+            # bound does not improve before switching to OA
+            max_nonimprove_iter = 5
+            making_progress = True
+            # TODO-romeo Unneccesary for OA and LOA, right?
+            for i in range(1, max_nonimprove_iter + 1):
+                try:
+                    if (sign_adjust * log[-i]
+                            <= (log[-i - 1] + progress_required)
+                            * sign_adjust):
+                        making_progress = False
+                    else:
+                        making_progress = True
+                        break
+                except IndexError:
+                    # Not enough history yet, keep going.
                     making_progress = True
                     break
-            except IndexError:
-                # Not enough history yet, keep going.
-                making_progress = True
-                break
-        if not making_progress and (
-                config.strategy == 'hPSC' and
-                config.strategy == 'PSC'):
-            config.logger.info(
-                'Not making enough progress for {} iterations. '
-                'Switching to OA.'.format(max_nonimprove_iter))
-            config.strategy = 'OA'
+            if not making_progress and (
+                    config.strategy == 'hPSC' and
+                    config.strategy == 'PSC'):
```

Comment:
There's a logical error in the `and` here.

Reply:
@qtothec is this logic error you previously pointed out still an issue?

Reply:
@blnicho Yes, good catch. I had changed it in my own copy of the code.